flink-runtime/src/main/java/org/apache/flink/runtime/client/JobClient.java

```diff
@@ -275,9 +275,35 @@ public static ClassLoader retrieveClassLoader(
 	public static JobExecutionResult awaitJobResult(JobListeningContext listeningContext) throws JobExecutionException {
 
 		final JobID jobID = listeningContext.jobID;
+		final ActorRef jobClientActor = listeningContext.jobClientActor;
 		final Future<Object> jobSubmissionFuture = listeningContext.jobResultFuture;
 		final ClassLoader classLoader = listeningContext.classLoader;
 
+		while (!jobSubmissionFuture.isCompleted()) {
+			try {
+				Thread.sleep(250);
```

Comment:
Can't we wait on the `jobSubmissionFuture` for, e.g., Flink's ask timeout and recover timeout exceptions by sending an `Identify` message to the JobClient actor. If he replies in time, then re-wait for the `jobSubmissionFuture`.